internal/template-validator/webhooks/hook.go

```diff
@@ -19,53 +19,147 @@
+	"kubevirt.io/ssp-operator/internal/template-validator/virtinformers"
 )
 
 const (
-	VMTemplateValidatePath string = "/virtualmachine-template-validate"
+	VmValidatePath       string = "/virtualmachine-validate"
+	TemplateValidatePath string = "/template-validate"
 )
 
-func ServeVMTemplateValidate(resp http.ResponseWriter, req *http.Request) {
-	serve(resp, req, admitVMTemplate)
+type admitFunc func(*admissionv1.AdmissionReview) *admissionv1.AdmissionResponse
+
+type Webhooks interface {
+	Register()
 }
 
-type admitFunc func(*admissionv1.AdmissionReview) *admissionv1.AdmissionResponse
+func NewWebhooks(apiClient client.Client, informers *virtinformers.Informers) Webhooks {
+	return &webhooks{
+		apiClient: apiClient,
+		informers: informers,
+	}
+}
+
+type webhooks struct {
+	apiClient client.Client
+	informers *virtinformers.Informers
+}
+
+func (w *webhooks) Register() {
+	http.HandleFunc(VmValidatePath, func(resp http.ResponseWriter, req *http.Request) {
+		serve(resp, req, w.admitVm)
+	})
+	http.HandleFunc(TemplateValidatePath, func(resp http.ResponseWriter, req *http.Request) {
+		serve(resp, req, w.admitTemplate)
+	})
+}
 
-func admitVMTemplate(ar *admissionv1.AdmissionReview) *admissionv1.AdmissionResponse {
-	newVM, oldVM, err := GetAdmissionReviewVM(ar)
+func (w *webhooks) admitVm(ar *admissionv1.AdmissionReview) *admissionv1.AdmissionResponse {
+	vm, err := GetAdmissionReviewVM(ar)
 	if err != nil {
 		return ToAdmissionResponseError(err)
 	}
 
-	if newVM.DeletionTimestamp != nil {
+	if vm.DeletionTimestamp != nil {
 		return ToAdmissionResponseOK()
 	}
 
-	rules, err := getValidationRulesForVM(newVM)
+	rules, err := getValidationRulesForVM(vm, w.informers.TemplateStore())
 	if err != nil {
 		return ToAdmissionResponseError(err)
 	}
 
-	log.Log.V(8).Infof("admission newVM:\n%s", spew.Sdump(newVM))
-	log.Log.V(8).Infof("admission oldVM:\n%s", spew.Sdump(oldVM))
+	log.Log.V(8).Infof("admission vm:\n%s", spew.Sdump(vm))
 	log.Log.V(8).Infof("admission rules:\n%s", spew.Sdump(rules))
 
-	causes := ValidateVMTemplate(rules, newVM, oldVM)
+	causes := ValidateVm(rules, vm)
 	if len(causes) > 0 {
 		return ToAdmissionResponse(causes)
 	}
 
 	return ToAdmissionResponseOK()
 }
 
+func (w *webhooks) admitTemplate(ar *admissionv1.AdmissionReview) *admissionv1.AdmissionResponse {
+	if ar.Request.Operation != admissionv1.Delete {
+		return ToAdmissionResponseOK()
+	}
+
+	template, err := GetAdmissionReviewTemplate(ar)
+	if err != nil {
+		return ToAdmissionResponseError(err)
+	}
+
+	// Check if template is a common template
+	_, ok := template.Labels[common_templates.TemplateTypeLabel]
+	if !ok {
+		return ToAdmissionResponseOK()
+	}
+
+	// Old versions of common templates had annotation validation
+	_, ok = template.Annotations[annotationValidationKey]
+	if !ok {
+		return ToAdmissionResponseOK()
+	}
+
+	// Old template cannot be removed if a VM uses it.
+	// We cannot filter by labels, because a VM may refer to the template by annotation.
```

Comment:
Shouldn't have all vms created from common template special labels, which points them to template? From CT release v0.6.0 should have all templates `vm.kubevirt.io/template` label.

Reply:
Yes, but I was not sure how to handle VMs from older templates.
I will change this part of the code. Calling List API in a webhook is probably too slow and may timeout if there are a lot of VMs.

Reply:
The older vms have `vm.cnv.io/template: <template name>` label, but you are right, that currently you can't put `OR` between two labels.
To not fetch too many results, you can use Limit variable in ListOptions and then update VMs in batches.

Reply:
I will post a different solution soon. We can use a lightweight controller, to watch all VMs, but only cache VMs that are relevant, so it will be memory efficient. Then this validating webhook can use the cache.